File: app/core/GameLoop.pl
:- consult(['./Gameplay.pl','../UI/Render.pl', '../util/Helpers.pl']).

/*
    Define o loop principal do Menu.
 Executa a tela de Menu enquanto recebe e valida as opções de entrada do player.
*/
menu_loop :-
    update_screen_state("menu"),
    action,
    read_line(Out),
    validation_input(["I", "R", "C", "S"], Out, ValidationOut),
    menu_resolve(ValidationOut).

/*
    Resolve o seguimento do loop para cada entrada dada pelo player.
*/
menu_resolve("I") :- !.

menu_resolve("R") :-
    update_screen_state("ranking"),
    ranking_loop, !.

menu_resolve("C") :-
    update_screen_state("creditos"),
    creditos_loop, !.

menu_resolve("S") :-
    halt.

menu_resolve(_) :-
    menu_loop.

/*
    Define o loop da tela de ranking.
 Executa o loop da tela de ranking até uma entrada válida do player.
*/
ranking_loop :-
    action,
    read_line(Out),
    validation_input(["V"], Out, ValidationOut),
    (atom_string(ValidationOut, "V") -> menu_loop; ranking_loop).
    %ranking_resolve(ValidationOut).

/*
    Define o loop da tela de créditos.
 Executa o loop da tela de créditos até uma entrada válida do player.
*/
creditos_loop :-
    action, 
    read_line(Out),
    validation_input(["V"], Out, ValidationOut),
    (atom_string(ValidationOut, "V") -> menu_loop; creditos_loop).

/*
    Define o loop que espera as informações do player para a campanha.
*/
desafiante_loop :-
    update_screen_state("desafiante"),
    action,
    read_line(Out),
    (atom_string(Out, "") -> desafiante_loop; init_campaign_state(Out)).

/*
    Define o loop da campanha do jogo.
 Define o loop para os casos de campanha em andamento, vitória do player, derrota
 do player para o bot, resolvendo os casos. 
*/
campaign_loop :-
    verify_victory(Out),
    (Out =:= 0),

    battle_loop,
    cards_comparation, % a atualização do score vem aqui, baseada na função comparationLoop em haskell.
    campaign_loop, !. 

campaign_loop :-
    verify_victory(Out),
    (Out =:= 1),

    get_player_state(PlayerData),
    get_campaign_state(CampaingData),

    nth0(0, PlayerData, PlayerScore),
    nth0(1, CampaignData, CampaignScore),
    nth0(2, CampaignData, CampaignLife),
    nth0(3, CampaignData, CampaignBelt),

    campaign_life_situation(Life),
    NextScore is PlayerScore + CampaignScore,
    NextBelt is CampaingBelt + 1,

    update_campaign_state(NextScore, Life, NextBelt),
    build_battle,
    (NextBelt =:= 6 -> campaign_win; campaign_loop), !.

campaign_loop :-
    verify_victory(Out),
    (Out =:= -1),

    get_player_state(PlayerData),
    get_campaign_state(CampaingData),

    nth0(0, PlayerData, PlayerScore),
    nth0(1, CampaignData, CampaignScore),
    nth0(2, CampaignData, CampaignLife),
    nth0(3, CampaignData, CampaignBelt),

    NextScore is PlayerScore + CampaignScore,
    Life is CampaignLife - 1,

    update_campaign_state(NextScore, Life, CampaignBelt),
    build_battle,
    (Life =< 0 -> campaign_loser; campaign_loop), !.

campaign_loop :-
    get_player_state(PlayerData),
    get_campaign_state(CampaignData),

    nth0(0, PlayerData, PlayerScore),
    nth0(1, CampaignData, CampaignScore),
    nth0(2, CampaignData, CampaignLife),
    nth0(3, CampaignData, CampaignBelt),

    campaign_life_situation(Life),
    NextScore is PlayerScore + CampaignScore,

    update_campaign_state(NextScore, Life, CampaignBelt),
    build_battle,
    campaign_loop.

/*
    Define o loop da batalha atual.
*/
battle_loop :-
    update_screen_state("batalha"),
    action,
    read_line(Out),
    validation_input(["1", "2", "3", "4", "5", "6", "7", "8", "D"], Out,  ValidationOut),

    battle_resolve(ValidationOut).

/*
    Resolve para as entradas válidas na seleção de uma carta na tela de batalha.
*/
battle_resolve(Input) :-
    member(Input, ["1", "2", "3", "4", "5"]),

    get_player_state(PlayerData),
    nth0(2, PlayerData, PlayerDeck),
    number_string(PlayerChoiceTemp, Input),
    PlayerChoice is PlayerChoiceTemp - 1,

    % bot_choice:
    get_bot_state(BotData),
    nth0(2, BotData, BotDeck),
    BotChoice = 0, % placeholder, cadê o Bot?

    play_card(PlayerDeck, PlayerChoice, NewPlayerDeck),
    play_card(BotDeck, BotChoice, NewBotDeck),

    sub_at(NewPlayerDeck, 2, PlayerData, NewPlayerData),
    sub_at(NewBotDeck, 2, BotData, NewBotData),

    update_player_state(NewPlayerData),
    update_bot_state(NewBotData), !.




% ==================== AUXILIARES ==================== %

/*
    Inicia os parâmetros iniciais de uma batalha.
*/
build_battle :-
    update_screen_state("batalha"),
    build_deck(PlayerDeck),
    build_deck(BotDeck),
    init_battle_state(PlayerDeck, BotDeck).
    
/*
    Constrói um deck de cartas permutada.
*/
build_deck(Deck) :-
    fill_deck(1, PureDeck),
    random_permutation(PureDeck, Deck).

/*
    Atualiza, se necessário, a vida do Player quando a pontuação for de 25.
*/
campaign_life_situation(Result) :-
    get_player_state(PlayerData),
    get_campaign_state(CampaingData),

    nth0(0, PlayerData, PlayerScore),
    nth0(2, CampaignData, CampaignLife),
    PossibleNewLife is CampaignLife + 1,

    (PlayerScore >= 25 -> Result = PossibleNewLife; Result = CampaignLife).


/*
    Lê a entrada do usuário de modo que não precise de um ponto no final da entrada.
*/
read_line(Result) :-
    read_line_to_codes(user_input, Codes),
    string_codes(Result, Codes).

/*
    Valida se uma entrada do usuário está de acordo com as opções atuais.
 Recebe uma lista de opções válidas, uma string que representa a entrada do player
 na situação atual e um parâmetro que representa a saída.
*/
validation_input(Options, Input, Result) :-
    string_chars(Input, InputToChar),
    contains_input(Options, InputToChar, OutContains),
    length(OutContains, Total),

    LastOut is Total - 1,
    (OutContains == [] -> Result = "";
    nth0(LastOut, OutContains, OutResult), Result = OutResult).

/*
    Retorna as entradas que estão dentre as válidas para o loop atual.
 Recebe uma lista com as opções válidas, uma lista contendo os caracteres digitados
 pelo jogador e o parâmetro de saída.
*/
contains_input(_, [], []) :- !.
contains_input(Options, [HInput | TInput], [Hresult | Tresult]) :-
    string_upper(HInput, InputCurrent),
    member(InputCurrent, Options),

    Hresult = InputCurrent,
    contains_input(Options, TInput, Tresult), !.

contains_input(Options, [_ | TInput], Result) :-
    contains_input(Options, TInput, Result).

File: haskell/app/core/GameLoop.hs
{-
    Módulo de execução do loop geral de jogo.
-}
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
module GameLoop where

import Bot
import Tips
import State
import Ranking
import Gameplay 
import StateManager
import Deck (playCard)
import Render (action)
import Helpers (organizeRankings)

import System.Process
import System.Info (os)
import System.Exit (die)
import Data.Char (toUpper)
import CSVManager (getCards)


-- | Inicia o loop principal do jogo e estabelece os dados iniciais. Após 
-- iniciado, chama a função que inicia o loop do menu.
initLoop :: IO()
initLoop = do
    let clear = if os == "mingw32" then "cls" else "clear"
    callCommand clear

    menuLoop

-- | Esta executa o loop sobre as possibilidades de escolha na tela de menu.
-- O loop é repetido até o jogador digitar um valor possível para escolha.
menuLoop :: IO()
menuLoop = do
    localUpdateScreen "menu"
    action
    input <- getLine

    menuResolve $ validInput ["I", "R", "C", "S"] input

-- | Esta função chama o próximo loop a ser executado de acordo com a escolha do
-- jogador. Caso não seja uma escolha válida, o loop atual é chamado mais uma vez.
menuResolve :: String -> IO()
menuResolve "I" = nameInput
menuResolve "R" = rankingLoop
menuResolve "C" = creditLoop
menuResolve "S" = do
    if os == "mingw32" then callCommand "cls" else callCommand "clear"
    die "Carta-Ninja Encerrado"
menuResolve _ = menuLoop

-- | Esta função executa o loop sobre as possibilidades de escolha na tela de ranking.
-- O loop é repetido até o jogador digitar um valor possível para escolha.
rankingLoop :: IO()
rankingLoop = do
    localUpdateScreen "ranking"
    action

    input <- getLine
    let choice = validInput ["V"] input
    if choice /= "" then menuLoop else rankingLoop

-- | Esta função executa o loop sobre as possibilidades de escolha na tela de créditos.
-- O loop é repetido até o jogador digitar um valor possível para escolha.
creditLoop :: IO()
creditLoop = do
    localUpdateScreen "creditos"
    action

    input <- getLine
    let choice = validInput ["V"] input
    if choice /= "" then menuLoop else creditLoop

-- | Esta função executa o loop sobre a requisição do nome do desafiante.
-- O loop é repetido até o jogador digitar um nome válido.
nameInput :: IO()
nameInput = do 
    localUpdateScreen "desafiante"
    resetCampaignState
    action 
    
    input <- getLine
    if input == "" 
        then nameInput 
        else do
            updatePlayerName input
    
            initBattleData
            battleLoop

-- | Esta função executa o loop sobre o estágio de batalha entre o jogador e o bot.
-- O loop é repetido até a partida encerrar em caso de vitória, de derrota, de empate ou 
-- caso o jogador não digite uma opção válida dentre as possíveis.
battleLoop :: IO()
battleLoop = do
    battle <- getBattleState

    let victoryCheck = verifyVictory battle

    if victoryCheck /= 0 then comparationResolve victoryCheck 
    else do 
        localUpdateScreen "batalha"
        action

        input <- getLine
        let playerDecide = validInput ["1", "2", "3", "4", "5", "6", "7", "8", "D"] input

        battleResolve battle playerDecide

        newBattleStage <- getBattleState
        let winner = 
                if verifyNullifyElemSpecialCardUse newBattleStage then 
                    getWinnerNullElement (last $ playerDeck newBattleStage) (last $ cpuDeck newBattleStage)
                else
                    getWinner (last $ playerDeck newBattleStage) (last $ cpuDeck newBattleStage)

        comparationLoop winner

-- | Esta função executa o loop sobre a escolha do player no estágio de batalha, permitindo
-- a ação das cartas especiais, das cartas comuns ou da dica.
battleResolve :: BattleState -> String -> IO()
battleResolve battle "6" = specialResolve battle 6
battleResolve battle "7" = specialResolve battle 7
battleResolve battle "8" = specialResolve battle 8
battleResolve battle "D" = tipResolve battle
battleResolve _ "" = battleLoop
battleResolve battle choice = do
    campaign <- getCampaignState

    let botChoice = makeChoice battle (beltLevel campaign)

    let newPlayerDeck = playCard (playerDeck battle) $ read choice - 1
    let newBotDeck = playCard (cpuDeck battle) botChoice

    let modifiedData = BattleState {
                currentRound = currentRound battle,

                playerScore = playerScore battle,
                playerStreak = playerStreak battle,
                playerWinsByElement = playerWinsByElement battle,
                playerDeck = newPlayerDeck,

                cpuScore = cpuScore battle,
                cpuStreak = cpuStreak battle,
                cpuWinsByElement = cpuWinsByElement battle,
                cpuDeck = newBotDeck,

                specialDeck = specialDeck battle,
                specialCardInUse = specialCardInUse battle,
                tipAvailability = tipAvailability battle
            }

    writeBattleState modifiedData

    if tipAvailability battle == "PRESSIONE [D] PARA USAR A DICA" then updateTipOf $ tipAvailability battle
    else updateTipOf "DICA USADA"

-- | Esta função resolve a utilização de uma carta especial, quando disponível.
specialResolve :: BattleState -> Int -> IO()
specialResolve battle specialNum = 
        if verifySpecialCardAvailability battle == 1 then do
            useSpecialCard specialNum
            battleLoop
        else battleLoop

-- | Esta função resolve a utilização da funcionalidade de dicas.
tipResolve :: BattleState -> IO()
tipResolve battle = do
    if tipAvailability battle == "PRESSIONE [D] PARA USAR A DICA" then do
        updateTipOf $ giveTip battle
        battleLoop
    else do
        battleLoop

-- | Esta função executa o loop sobre o estágio de comparação das cartas.
-- Caso haja definição na batalha, o próximo estágio declarado será o de rosolução
-- de uma comparação.
comparationLoop :: Int -> IO()
comparationLoop winner = do
    battle <- getBattleState
    localUpdateScreen "comparacao"
    action
    
    if winner == 1 then 
        updateScoreOf winner (last $ playerDeck battle)
    else
        updateScoreOf winner (last $ cpuDeck battle)

    _ <- getLine

    battleLoop

-- | Esta função resolve a situação do jogador para a batalha.
comparationResolve :: Int -> IO()
comparationResolve (-2) = battleDrawLoop
comparationResolve (-1) = battleDefeatLoop
comparationResolve 1 = battleVictoryLoop

-- | Esta função executa o loop sobre o estágio de empate em uma partida.
-- Nesse caso, a partida é repetida sem problemas.
battleDrawLoop :: IO()
battleDrawLoop = do
    battle <- getBattleState
    updatePlayerCampaignScore $ playerScore battle

    localUpdateScreen "empate"
    action

    _ <- getLine

    initBattleData
    battleLoop

-- | Esta função executa o loop sobre o estágio de derrota em uma partida.
-- Nesse caso, a partida é repetida, caso o player possua vidas restantes;
-- a campanha é encerrada, caso o player não tenha vidas restantes.
battleDefeatLoop :: IO()
battleDefeatLoop = do
    battle <- getBattleState
    campaign <- getCampaignState
    updatePlayerCampaignScore $ playerScore battle

    if lifes campaign > 0 then do
        localUpdateScreen "derrota"
        action
        _ <- getLine
        initBattleData
        updatePlayerLife (-1)
        battleLoop
    else do 
        localUpdateScreen "gameOver"
        action
        _ <- getLine
        updateRanking
        rankingLoop

-- | Esta função executa o loop sobre o estágio de vitória em uma partida.
-- Nesse caso, o jogador avança de nível, caso ele esteja em até um nível abaixo
-- do último; o jogo encerra, caso todos os níveis tiverem sido jogados.
battleVictoryLoop :: IO()
battleVictoryLoop = do
    battle <- getBattleState
    campaign <- getCampaignState
    updatePlayerCampaignScore $ playerScore battle

    if playerScore battle >= 25 then updatePlayerLife 1 else updatePlayerLife 0

    if beltLevel campaign >= 5 then do
        localUpdateScreen "gameClear"
        action
        _ <- getLine
        updateRanking
        rankingLoop
    else do
        localUpdateScreen "vitoria"
        action
        _ <- getLine
        levelUpPlayer
        initBattleData
        battleLoop

------------------------------ Funções Auxiliares ------------------------------

-- | Inicia os dados de um estágio de batalha utilizando um padrão comum.
initBattleData :: IO()
initBattleData = do
    let deckDefault = getCards
    playerRandomDeck <- shuffle deckDefault
    cpuRandomDeck <- shuffle deckDefault

    let modifiedData = BattleState {
                currentRound = 1,

                playerScore = 0,
                playerStreak = 0,
                playerWinsByElement = [False, False, False, False, False],
                playerDeck = playerRandomDeck,

                cpuScore = 0,
                cpuStreak = 0,
                cpuWinsByElement = [False, False, False, False, False],
                cpuDeck = cpuRandomDeck,

                specialDeck = ["swapInDeck", "nullifyElement", "swapBetweenDecks"],
                specialCardInUse = False,
                tipAvailability = "PRESSIONE [D] PARA USAR A DICA"
            }

    writeBattleState modifiedData

-- | Adiciona o ranking finalizado do jogador atual.
updateRanking :: IO()
updateRanking = do
    global <- getGlobalState
    campaign <- getCampaignState

    let newRanking = Ranking {points = totalScore campaign, name = playerName campaign}

    let modifiedData = 
            GlobalState {
                screen = screen global, 
                rankings = organizeRankings $ newRanking : rankings global
                }
    writeGlobalState modifiedData

-- | Função que denota a nova tela que representa o loop atual de jogo. Recebe 
-- um argumento que representa a nova tela a ser escrita. 
localUpdateScreen :: String -> IO()
localUpdateScreen newScreen = do
    currentGlobals <- getGlobalState
    writeGlobalState $ GlobalState {screen = newScreen, rankings = rankings currentGlobals}

-- | A partir de uma lista de entradas válidas para a tela atual e uma String que
-- é a entrada, esta função retorna a última entrada válida digitada pelo player
-- antes de um [enter]. Palavra vazia se não houver entrada válida.
validInput :: [String] -> String -> String
validInput validKeys input = do
    let inputsPossible = filter (`elem` validKeys) $ splitInput input

    if not (null inputsPossible) then last inputsPossible else ""

-- | Separa os elementos de uma String e os coloca em letra caixa alto, isso
-- quando possível.
splitInput :: String -> [String]
splitInput "" = []
splitInput (h:t) = [toUpper h] : splitInput t